### PR TITLE
Polariton qc

### PR DESCRIPTION
**Pull request by FoleyLab**
## Description
This pull request provides reference implementations of hybrid ab initio / cavity quantum electrodynamics methods for computing mean-field ground- and excited-states of molecules strongly coupled to a single quantized photon. 

## What are your new additions? Please provide a brief list.
* **New Features**
  - [x] Restricted Hartree-Fock energy and orbitals for a molecule coupled to a single quantized photon provided in helper_CQED_RHF.py
  - [x] Configuration Interaction Singles energies and wavefunctions for a molecule coupled to a single quantized photon in helper_CS_CQED_CIS.py
  - [x] Ability to provide complex energy for the photon to account for dissipation / energy uncertainty in the photonic degree of freedom is provided in helper_CS_CQED_CIS.py
  - [x] Example calculations with tests are provided in CQED_RHF.py and CS_CQED_CIS.py
  - [x] Key equations and references provided in README.md  

## Any questions for the community?
- [ ] The CS_CQED_CIS has not been reported in the literature to my knowledge, so I don't know of peer-reviewed reference values, so the main independent test comes from the limit where the molecule is not coupled to a photon, where the method reduces to ordinary CIS which can be compared directly with Psi4 values - is that adequate?

## Status
- [x] Click when ready for review-and-merge

### CODE REVIEW

**Review by loriab — COMMENTED** [review]
Very nice reference implementation, thank you!

After clarification on whether some of the outside-of-Polaritonic-directory changes were intentional, lgtm.

File: Polaritonic-Quantum-Chemistry/CS_CQED_CIS.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+# collect every other excitation energy for case 1 since cqed-cis will be doubly
+# degenerate when omega = 0 and lambda = 0,0,0
+cqed_cis_e_1 = cqed_cis_e_1[::2]
+
+print("\n    PRINTING RESULTS FOR CASE 1: NO COUPLING")
+print("\n    CASE 1 RHF Energy (Hatree):                    %.8f" % psi4_rhf_e)
+print("    CASE 1 CQED-RHF Energy (Hartree):              %.8f" % scf_e_1)
+print(
+    "    CASE 1 CIS LOWEST EXCITATION ENERGY (eV)        %.4f"
+    % (psi4_excitation_e[0] * 27.211)
```
Comment (dgasmith) [inline Polaritonic-Quantum-Chemistry/CS_CQED_CIS.py:106]: Use `psi4.constants...` @loriab ?

File: Polaritonic-Quantum-Chemistry/helper_CS_CQED_CIS.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+    # elements corresponding to <s|<\Phi_0 | H | \Phi_i^a>|t>
+    # Eq. (17) of [McTague:2021:ChemRxiv]
+    for s in range(0, 2):
+        for i in range(0, ndocc):
+            for a in range(0, nvirt):
+                A = a + ndocc
+                for t in range(0, 2):
+                    iat = 2 * (i * nvirt + a) + t + 2
+                    Hep[s, iat] = (
```
Comment (dgasmith) [inline Polaritonic-Quantum-Chemistry/helper_CS_CQED_CIS.py:186]:
```suggestion
                    Hep[s, iat] = Hep[iat, s] = (
```

We shouldn't duplicate the transpose.

File: Polaritonic-Quantum-Chemistry/helper_CS_CQED_CIS.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+    Hp = np.zeros((ndocc * nvirt * 2 + 2, ndocc * nvirt * 2 + 2), dtype=complex)
+
+    # elements corresponding to <s|<\Phi_0 | H | \Phi_0>|t>
+    # Eq. (16) of [McTague:2021:ChemRxiv]
+    Hp[0, 0] = 0.0
+    Hp[1, 1] = omega_val
+
+    # elements corresponding to <s|<\Phi_0 | H | \Phi_i^a>|t>
+    # Eq. (17) of [McTague:2021:ChemRxiv]
+    for s in range(0, 2):
```
Comment (dgasmith) [inline Polaritonic-Quantum-Chemistry/helper_CS_CQED_CIS.py:180]: Possible to use broadcasting here and build it up in layers.

**Review by dgasmith — COMMENTED** [review]
I cannot comment too much on the code, but the overall layout looks good to me.

Anyone here familiar with the subject or perhaps someone in the CC/CI space to review?

File: Tutorials/03_Hartree-Fock/3a_restricted-hartree-fock.ipynb [hunk 4]
```diff
@@ -233,18 +433,21 @@
     "# Memory check for ERI tensor\n",
     "I_size = (nbf**4) * 8.e-9\n",
     "print('\\nSize of the ERI tensor will be {:4.2f} GB.'.format(I_size))\n",
+    "memory_footprint = I_size * 1.5\n",
     "if I_size > numpy_memory:\n",
     "    psi4.core.clean()\n",
     "    raise Exception(\"Estimated memory utilization (%4.2f GB) exceeds allotted memory \\\n",
-    "                     limit of %4.2f GB.\" % (I_size, numpy_memory))\n",
+    "                     limit of %4.2f GB.\" % (memory_footprint, numpy_memory))\n",
```
[the remark this answers is not in the document]
Comment (FoleyLab) [reply Tutorials/03_Hartree-Fock/3a_restricted-hartree-fock.ipynb:440]: Ah yes, totally incidental that I did not mean to commit!

File: Polaritonic-Quantum-Chemistry/README.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+The following codes are available:
+- `CQED_RHF.py`: A script to compute a modified Restricted Hartree-Fock ground-state of the water molecule strongly coupled to a single photon.
+- `CS_CQED_CIS.py`: A script to compute the ground-state and excitation energies of a lone MgH+ molecule (Case 1), MgH+ strongly 
+                    coupled to a single photon that is resonant with the first dipole-allowed transition (Case 2), and
+                    MgH+ coupled to a single photon with finite lifetime / complex energy with central energy resonant with the first dipole
+                    allowed-transition (Case 3).
+
+Helper programs:
+- `helper_CQED_RHF.py`: A helper function to perform Restricted Hartree-Fock theory for the mean-field ground state of the  Pauli-Fierz Hamiltonian that includes an *ab initio* electronic Hamiltonian with dipolar coupling to a quantized photon mode and a quadratic self polarization energy contribution.  This implementation builds upon Psi4Numpy's [RHF.py](https://github.com/FoleyLab/psi4numpy/blob/polariton-qc/Self-Consistent-Field/RHF.py).
```
[the remark this answers is not in the document]
Comment (FoleyLab) [reply Polaritonic-Quantum-Chemistry/README.md:14]: Yes it should!

File: Polaritonic-Quantum-Chemistry/helper_CS_CQED_CIS.py (hunk 3 above)
Comment (FoleyLab) [reply Polaritonic-Quantum-Chemistry/helper_CS_CQED_CIS.py:180]: Thank you @dgasmith I will think about this - if there are any examples of building up matrices in this way in other reference implementations here lmk!

**Pull request closed without merging**